contracts/DKVDaggerHashimoto.sol

```diff
@@ -144,20 +147,33 @@ contract DecentralizedKVDaggerHashimoto is DecentralizedKV {
 
     /* ON CHAIN VERFICATION */
     function checkDaggerDataWithProof(
-        uint256 idx,
-        bytes32 kvHash,
+        uint256 chunkIdx,
+        PhyAddr memory kvInfo,
         bytes32[] memory proofs,
-        bytes memory maskedData
-    ) public pure returns (bool) {
-        bytes32 dataHash = keccak256(maskedData);
-        bytes32 rootFromProofs = MerkleLib.calculateRootWithProof(dataHash, idx, proofs);
+        bytes memory unmaskedData
+    ) public view returns (bool) {
+        uint256 chunksNumPerKV = 1 << chunkLenBits;
+        uint256 maxChunkIdx = lastKvIdx * chunksNumPerKV-1;
+
+        if (chunkIdx > maxChunkIdx){
+            return keccak256(unmaskedData) == EMPTY_CHUNK_4KB_HASH;
+        }
+        
+        uint256 chunkLeafIdx = chunkIdx % (1 << chunkLenBits);
+
+        if (chunkLeafIdx >=  MerkleLib.getMaxLeafsNum(kvInfo.kvSize,chunkSize)){
+            return keccak256(unmaskedData) == EMPTY_CHUNK_4KB_HASH;
+        }
```

Comment:
I saw Qiang's comments on these two edge cases. In principle, I appreciate the explicit expression on thoses two overflow cases.

Originally the reason I didn't do any restriction here is, a wrong index won't lead to a correct Merkle tree root calculation, which means in the end, it cannot return a false positive judgement on `bytes24(rootFromProofs) == bytes24(kvInfo.hash)`. Hereby I didn't explicitly show those two edge cases, to some extent it could save some gas.

But I do omit a case where unmaskedData indicates a empty hash chunk. In this case, chunkIdx is meaningless, a undefined behavior. My proposed version:

```
bytes32 dataHash = keccak256(unmaskedData);
if (dataHash == EMPTY_CHUNK_4KB_HASH) {
  return True;
}
bytes32 rootFromProofs = MerkleLib.calculateRootWithProof(dataHash, chunkLeafIdx, proofs);
return bytes24(rootFromProofs) == bytes24(kvInfo.hash);
```

But I am okay to the idea explicitly showing overflow cases. Up to you.

Reply:
I think the code above will lead to the chunk stored by the user being full with zero can escape the Merkle proving the process.

Reply:
You are right. 

Then how can we handle a care where user rent a space but for zero bytes? In this scenario `kvInfo.hash = 0x0` and kvIdx is not in use? 

How about this:
```
bytes32 dataHash = keccak256(unmaskedData);
if (dataHash == EMPTY_CHUNK_4KB_HASH && kvInfo.hash == bytes24(0x0)) {
  return True;
}
bytes32 rootFromProofs = MerkleLib.calculateRootWithProof(dataHash, chunkLeafIdx, proofs);
return bytes24(rootFromProofs) == bytes24(kvInfo.hash);
```

Is that enough?

Reply:
In short here are two separate problems:

1. How to deal with idx overflow
2. How to deal with actual no data stored (or say, a null kvInfo.hash from user)

Personally, I am questioning whether the first item has to be protected in this piece of code.
For second, it is indeed a problem and above lists my solution. 

Hope I clarify a bit better : )

Reply:
> if (dataHash == EMPTY_CHUNK_4KB_HASH && kvInfo.hash == bytes24(0x0)) {
>   return True;
> }

seems like still do not handle the second case: "the key-value hash is minTree hash, so if the sample trunk index is large than the key-value size, then we should not verify it by the proof but also check if the trunk is 0."

Reply:
What kind of user data could lead to this case, if I may?

Reply:
ChunkSize = 4k KVSize = 16K  (4 chunks per kvSIze)
The user stores 8KB data of kvIdx 0 .
The minTree will only include the first two leaves and the last two leaves are empty.